RIOT-OS/RIOT
# cpu/nrfx: rework periph_uart driver to allow use of multiple UARTs with nrf52840

### Contribution description

The initial goal of this PR was to update the UART configuration of the nrf52840dk board with a second UART, since there are 2 available on this CPU. I need this for #10563 where external radios devices (for WiFi, NB_IoT and maybe GSM) uses the second UART with AT commands.

When working on this, I realized the second UART needs a different struct defined in the CMSIS (NRF_UART**E**_TYPE) so it couldn't be added as-is. But the good thing is that the first UART can also use this other struct. Then the only missing thing was adapting the `periph_uart` driver to this new struct: it requires to use EasyDMA for transfering data.

So, to summarize, this PR provides several things:
- adapts the `periph_uart` to be compatible with NRF_UARTE_TYPE, only used with nrf52840xxx cpus
- provides hardware flow control with nrf52840 cpus and configure STDIO to use this on nrf52840dk
- use the array style configuration for UART config with nrf52840dk board (or nrf52840 based boards in general)
- configure the second UART available on the Arduino pinout of this board.

This PR was tested with success using the `tests/periph_uart` application.

There is also a small increase of ROM size (+36 bytes) for nrf51/nrf52 based boards. The ROM size increase is more important with nrf52840dk, (+190 bytes), probably because there's a second UART and also some parts are now very different from the initial driver.
Note that I'm not an expert in ROM size optimization, so any help or advice is welcome.

Also note that this PR is based on #10279 for simplicity (because common files are modified/moved by both PRs).




### Testing procedure

If you don't have all the hardware, you can use the boards provided by IoT-LAB to verify everything is still working:

- Start an experiment on IoT-LAB
```
$ iotlab-experiment submit -n pr -d 60 -l 1,site=saclay+archi=nrf51dk:ble -l 1,site=saclay+archi=nrf52dk:ble -l 1,site=saclay+archi=nrf52840dk:multi
```
- Flash any application using the shell on the boards:
```
$ make BOARD=nrf51dk IOTLAB_NODE=auto-ssh -C examples/saul flash term
$ make BOARD=nrf52dk IOTLAB_NODE=auto-ssh -C examples/saul flash term
$ make BOARD=nrf52840dk IOTLAB_NODE=auto-ssh -C examples/saul flash term
```
- Verify the shell works with all of them. This will validate that the existing driver (nrf51dk and nrf52dk) still works but also that the UART with EasyDMA is working as well (nrf52840dk).

Regarding the test of the second UART configured on the nrf52840dk, there are 2 possibilities:
1. Physically grab on this board, flash the `tests/periph_uart` application on it, shortcut RX/TX (D0/D1) pins, initialize the second UART and send data:
```
$ make BOARD=nrf52840dk -C tests/periph_uart flash term
> init 1 115200
> send test
```
(You can also try different baudrates if you want)
2. Trust me that it also works ;)




### Issues/PRs references

This PR is based on #10279 and will be required by #9833 and #10563 very soon.
